test/UnitTestFramework/UnitTestFramework.js

```diff
@@ -167,8 +191,120 @@ var testRunner = function testRunner() {
+            }
+
+            Promise.all(asyncTest.promise.map(Promise.all, Promise)).then( function() {
+                passedTestCount = 0;
+                for (var i = 0; i < asyncTest.count.length; i++) {
+                    var passed = testRunner.asyncTestHasNoErr(i);
+                    if (_verbose) {
+                        helpers.writeln("*** Result of test #", i + 1, " (", i, "): ", testsToRun[i].name);
+                        if (!passed) helpers.writeln(asyncTest.error[i]);
+                        helpers.writeln(passed ? "PASSED" : "FAILED");
+                    }
+                    !passed || passedTestCount++;
+                }
+                if (_verbose || executedTestCount != passedTestCount) {
+                    helpers.writeln("Summary of tests: total executed: ", executedTestCount,
+                      "; passed: ", passedTestCount, "; failed: ", executedTestCount - passedTestCount);
+                } else {
+                    helpers.writeln("pass");
+                }
+            });
+        },
+
+        runAsyncTest: function runAsyncTest(testIndex, testName, testBody) {
+            ///<summary>Runs test body catching all exceptions.<br/>
+            /// Result: prints LOADED/LOADING FAILED to the output.
+            ///</summary>
+            function logTestNameIf(b) {
+                if (b) {
+                    helpers.writeln("*** Loading test #", executedTestCount + 1, " (", testIndex, "): ", testName);
+                }
+            }
+
+            logTestNameIf(_verbose);
+
+            asyncTest.testIndex = testIndex;
+            asyncTest.promise[testIndex] = [];
+            asyncTest.resolve[testIndex] = [];
+
+            var isSuccess = true;
+            try {
+                testBody(testIndex);
+            } catch (ex) {
+                var message = ex.stack || ex.message || ex;
+                logTestNameIf(!_verbose);
+                helpers.writeln("Test threw exception: ", message);
+                isSuccess = false;
+            }
+            if (isSuccess) {
+                if (_verbose) {
+                    helpers.writeln("LOADED");
+                }
+                ++passedTestCount;
+            } else {
+                helpers.writeln("LOADING FAILED");
+                testRunner.asyncTestErr(testIndex, "LOADING FAILED");
+            }
+            ++executedTestCount;
+        },
+
+        asyncTestBegin : function asyncTestBegin(testIndex, testCount) {
+            _asyncMode = true;
+            asyncTest.testIndex = testIndex;
+        },
+
+        asyncTestEnd : function asyncTestEnd(testIndex, testCount) {
+            _asyncMode = false;
+            asyncTest.resolve[testIndex][testCount]();
+        },
+
+        async : function async(source, shouldFail) {
+            var testIndex = asyncTest.testIndex;
+            if (!shouldFail) {
+                var testCount = asyncTest.count[testIndex]++;

```

Comment:
>var testCount = asyncTest.count[testIndex]++; [](start = 16, length = 45)

Trying to get count before or after increment? Should testIndex ever be the same? I don't quite understand the intent of this pattern